scrapscript.py

```diff
@@ -1359,8 +1292,149 @@ def bdecode(msg: str) -> object:
+                obj, op=obj.op.name.encode("utf-8"), left=self.serialize(obj.left), right=self.serialize(obj.right)
+            )
+        if isinstance(obj, EnvObject):
+            return self._serialize(obj, env=self.serialize_env(obj.env))
+        if isinstance(obj, List):
+            return self._serialize(obj, items=[self.serialize(item) for item in obj.items])
+        if isinstance(obj, MatchFunction):
+            return self._serialize(obj, cases=[self.serialize(case) for case in obj.cases])
+        if isinstance(obj, NativeFunction):
+            return self.serialize(NativeFunctionRelocation(obj.name))
+        if isinstance(obj, Closure):
+            return self._serialize(obj, env=self.serialize_env(obj.env), func=self.serialize(obj.func))
+        if isinstance(obj, Record):
+            return self._serialize(
+                obj, data={key.encode("utf-8"): self.serialize(value) for key, value in obj.data.items()}
+            )
+        if isinstance(obj, Assign):
+            return self._serialize(obj, name=self.serialize(obj.name), value=self.serialize(obj.value))
+        if isinstance(obj, Function):
+            return self._serialize(obj, arg=self.serialize(obj.arg), body=self.serialize(obj.body))
+        if isinstance(obj, MatchFunction):
+            return self._serialize(obj, cases=[self.serialize(case) for case in obj.cases])
+        if isinstance(obj, MatchCase):
+            return self._serialize(obj, pattern=self.serialize(obj.pattern), body=self.serialize(obj.body))
+        if isinstance(obj, Apply):
+            return self._serialize(obj, func=self.serialize(obj.func), arg=self.serialize(obj.arg))
+        raise NotImplementedError(f"serialization for {type(obj).__name__} is not supported")
+
+    def serialize_env(self, env: Env) -> Dict[bytes, object]:
+        # We assume Env is not recursive.
+        return {key.encode("utf-8"): self.serialize(value) for key, value in env.items()}
+
+
+@dataclass(frozen=True)
+class Link(Object):
+    idx: int
+    obj: typing.List[Object] = dataclasses.field(
+        default_factory=lambda: []
+    )  # Used as a cell without modifying the attribute
+
+    def link(self, obj: Object) -> None:
+        assert len(self.obj) == 0
+        self.obj.append(obj)
+
+
+class Deserializer:
+    def __init__(self, refs: typing.List[Dict[str, Any]]) -> None:
+        self.refs = refs
+        self.objs: typing.List[Object] = [Link(idx) for idx in range(len(refs))]
+
+    def deserialize_all(self) -> Object:
+        for idx in reversed(range(len(self.refs))):
+            obj = self._deserialize(self.refs[idx])
+            link = self.objs[idx]
+            assert isinstance(link, Link)
+            link.link(obj)
+            self.objs[idx] = obj
+            self.link(idx)
+        return self.objs[0]
+
+    # TODO(max): Figure out a way to link without having two entire functions
+    # that have knowledge about each object type
+    def link(self, idx: int) -> None:
```

Comment:
@gregorybchris have any ideas on how to do this? i am using one level of indirection but it's a little gross

Reply:
in particular the gross thing is having two tree traversals. if we can get it to one traversal, that would be fine imo

Reply:
No good ideas for you...